File: Frameworks/react-web/tests/unit/models/ods-app.test.ts
import { describe, it, expect } from 'vitest'
import { parseApp, type OdsApp } from '../../../src/models/ods-app.ts'

// ===========================================================================
// OdsApp model tests
// ===========================================================================

describe('parseApp', () => {
  // -------------------------------------------------------------------------
  // Minimal valid input
  // -------------------------------------------------------------------------

  it('parses minimal valid input with appName, startPage string, and pages', () => {
    const result = parseApp({
      appName: 'My App',
      startPage: 'homePage',
      pages: { homePage: { title: 'Home', components: [] } },
    })
    expect(result.appName).toBe('My App')
    expect(result.startPage).toBe('homePage')
    expect(result.pages).toHaveProperty('homePage')
  })

  // -------------------------------------------------------------------------
  // Role-based startPage
  // -------------------------------------------------------------------------

  it('resolves role-based startPage object to default value', () => {
    const result = parseApp({
      appName: 'RoleApp',
      startPage: { default: 'page1', admin: 'page2' },
      pages: {},
    })
    expect(result.startPage).toBe('page1')
  })

  it('builds startPageByRole from role-based startPage object (excludes default)', () => {
    const result = parseApp({
      appName: 'RoleApp',
      startPage: { default: 'page1', admin: 'page2' },
      pages: {},
    })
    expect(result.startPageByRole).toEqual({ admin: 'page2' })
  })

  it('returns empty map for string startPage', () => {
    const result = parseApp({
      appName: 'App',
      startPage: 'landing',
      pages: {},
    })
    expect(result.startPageByRole).toEqual({})
  })

  it('returns empty map for null startPage', () => {
    const result = parseApp({
      appName: 'App',
      startPage: null,
      pages: {},
    })
    expect(result.startPageByRole).toEqual({})
  })

  it('returns empty map for undefined startPage', () => {
    const result = parseApp({
      appName: 'App',
      pages: {},
    })
    expect(result.startPageByRole).toEqual({})
  })

  it('returns empty string for startPage when startPage is null', () => {
    const result = parseApp({
      appName: 'App',
      startPage: null,
      pages: {},
    })
    expect(result.startPage).toBe('')
  })

  it('returns empty string for startPage when startPage is undefined', () => {
    const result = parseApp({
      appName: 'App',
      pages: {},
    })
    expect(result.startPage).toBe('')
  })

  // -------------------------------------------------------------------------
  // Missing optional fields default to empty arrays/objects
  // -------------------------------------------------------------------------

  it('defaults menu to empty array when not provided', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.menu).toEqual([])
  })

  it('defaults dataSources to empty object when not provided', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.dataSources).toEqual({})
  })

  it('defaults settings to empty object when not provided', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.settings).toEqual({})
  })

  it('defaults tour to empty array when not provided', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.tour).toEqual([])
  })

  it('defaults pages to empty object when not provided', () => {
    const result = parseApp({ appName: 'App', startPage: 'p' })
    expect(result.pages).toEqual({})
  })

  // -------------------------------------------------------------------------
  // Full spec with all fields
  // -------------------------------------------------------------------------

  it('parses full spec including menu, dataSources, settings, help, tour, auth, branding', () => {
    const result = parseApp({
      appName: 'Full App',
      startPage: 'dashboard',
      pages: {
        dashboard: { title: 'Dashboard', components: [] },
      },
      menu: [
        { label: 'Home', target: 'dashboard' },
      ],
      dataSources: {
        tasks: { url: 'local://tasks', method: 'GET' },
      },
      settings: {
        theme: { key: 'theme', label: 'Theme', type: 'text' },
      },
      help: {
        url: 'https://help.example.com',
      },
      tour: [
        { target: '#welcome', content: 'Welcome!' },
      ],
      auth: {
        provider: 'oauth2',
        roles: ['admin', 'user'],
      },
      branding: {
        theme: 'nord',
        mode: 'dark',
      },
    })
    expect(result.appName).toBe('Full App')
    expect(result.startPage).toBe('dashboard')
    expect(result.menu).toHaveLength(1)
    expect(result.dataSources).toHaveProperty('tasks')
    expect(result.settings).toHaveProperty('theme')
    expect(result.tour).toHaveLength(1)
    expect(result.branding.theme).toBe('nord')
    expect(result.branding.mode).toBe('dark')
  })

  // -------------------------------------------------------------------------
  // Auth and branding defaults
  // -------------------------------------------------------------------------

  it('provides default auth when not specified', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.auth).toBeDefined()
  })

  it('provides default branding when not specified', () => {
    const result = parseApp({ appName: 'App', startPage: 'p', pages: {} })
    expect(result.branding).toBeDefined()
    expect(result.branding.theme).toBe('indigo')
    expect(result.branding.mode).toBe('system')
  })

  // -------------------------------------------------------------------------
  // startPageByRole filters non-string values
  // -------------------------------------------------------------------------

  it('filters non-string values and default from role-based startPage map', () => {
    const result = parseApp({
      appName: 'App',
      startPage: { default: 'page1', bad: 42, also_bad: null },
      pages: {},
    })
    expect(result.startPageByRole).toEqual({})
  })
})

File: Frameworks/flutter-local/test/models/ods_app_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/models/ods_app.dart';

void main() {
  // =========================================================================
  // Minimal valid app
  // =========================================================================

  group('OdsApp.fromJson minimal', () {
    test('parses appName, startPage string, and pages', () {
      final app = OdsApp.fromJson({
        'appName': 'Test App',
        'startPage': 'home',
        'pages': {
          'home': {
            'title': 'Home',
            'content': [],
          },
        },
      });
      expect(app.appName, 'Test App');
      expect(app.startPage, 'home');
      expect(app.pages.containsKey('home'), true);
      expect(app.pages['home']!.title, 'Home');
    });

    test('startPageByRole is empty for string startPage', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.startPageByRole, isEmpty);
    });
  });

  // =========================================================================
  // Role-based startPage
  // =========================================================================

  group('OdsApp.fromJson role-based startPage', () {
    test('resolves startPage to default from object', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': {
          'default': 'page1',
          'admin': 'page2',
        },
        'pages': {
          'page1': {'title': 'Page 1', 'content': []},
          'page2': {'title': 'Page 2', 'content': []},
        },
      });
      expect(app.startPage, 'page1');
    });

    test('startPageByRole has role entries (without default)', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': {
          'default': 'page1',
          'admin': 'page2',
        },
        'pages': {
          'page1': {'title': 'Page 1', 'content': []},
          'page2': {'title': 'Page 2', 'content': []},
        },
      });
      expect(app.startPageByRole['admin'], 'page2');
      expect(app.startPageByRole.containsKey('default'), false);
    });
  });

  // =========================================================================
  // Missing optional fields default to empty
  // =========================================================================

  group('OdsApp.fromJson optional fields', () {
    test('menu defaults to empty list', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.menu, isEmpty);
    });

    test('dataSources defaults to empty map', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.dataSources, isEmpty);
    });

    test('settings defaults to empty map', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.settings, isEmpty);
    });

    test('tour defaults to empty list', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.tour, isEmpty);
    });

    test('help defaults to null', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.help, isNull);
    });
  });

  // =========================================================================
  // Full spec with all fields
  // =========================================================================

  group('OdsApp.fromJson full spec', () {
    test('parses all top-level fields', () {
      final app = OdsApp.fromJson({
        'appName': 'Full App',
        'startPage': 'dashboard',
        'menu': [
          {'label': 'Dashboard', 'mapsTo': 'dashboard'},
          {'label': 'Settings', 'mapsTo': 'settings'},
        ],
        'pages': {
          'dashboard': {
            'title': 'Dashboard',
            'content': [
              {'component': 'text', 'content': 'Welcome'},
            ],
          },
          'settings': {
            'title': 'Settings',
            'content': [],
          },
        },
        'dataSources': {
          'tasks': {
            'method': 'GET',
            'url': 'local://tasks',
          },
        },
        'help': {
          'overview': 'This is a test app.',
          'pages': {'dashboard': 'Dashboard help text'},
        },
        'tour': [
          {'title': 'Welcome', 'content': 'Start here', 'page': 'dashboard'},
        ],
        'settings': {
          'currency': {
            'label': 'Currency',
            'type': 'select',
            'default': 'USD',
            'options': ['USD', 'EUR'],
          },
        },
      });
      expect(app.appName, 'Full App');
      expect(app.startPage, 'dashboard');
      expect(app.menu.length, 2);
      expect(app.menu[0].label, 'Dashboard');
      expect(app.pages.length, 2);
      expect(app.help, isNotNull);
      expect(app.help!.overview, 'This is a test app.');
      expect(app.help!.pages['dashboard'], 'Dashboard help text');
      expect(app.tour.length, 1);
      expect(app.tour[0].title, 'Welcome');
      expect(app.settings.containsKey('currency'), true);
    });
  });

  // =========================================================================
  // startPageForRoles()
  // =========================================================================

  group('startPageForRoles', () {
    late OdsApp app;

    setUp(() {
      app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': {
          'default': 'home',
          'admin': 'adminDash',
          'editor': 'editorDash',
        },
        'pages': {
          'home': {'title': 'Home', 'content': []},
          'adminDash': {'title': 'Admin', 'content': []},
          'editorDash': {'title': 'Editor', 'content': []},
        },
      });
    });

    test('resolves admin role to admin page', () {
      expect(app.startPageForRoles(['admin']), 'adminDash');
    });

    test('resolves editor role to editor page', () {
      expect(app.startPageForRoles(['editor']), 'editorDash');
    });

    test('falls back to default when no role matches', () {
      expect(app.startPageForRoles(['viewer']), 'home');
    });

    test('falls back to default for empty roles list', () {
      expect(app.startPageForRoles([]), 'home');
    });

    test('returns first matching role when user has multiple roles', () {
      expect(app.startPageForRoles(['editor', 'admin']), 'editorDash');
    });
  });

  // =========================================================================
  // startPageByRole edge case
  // =========================================================================

  group('startPageByRole', () {
    test('returns empty map for string startPage', () {
      final app = OdsApp.fromJson({
        'appName': 'App',
        'startPage': 'home',
        'pages': {
          'home': {'title': 'Home', 'content': []},
        },
      });
      expect(app.startPageByRole, isEmpty);
    });
  });
}
